### PR TITLE
bridged-tokens-wrapper SC changes

### PR DESCRIPTION
**Pull request by alyn509**

### CODE REVIEW

**Review by dragos-rebegea — APPROVED** [review]

File: bridged-tokens-wrapper/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+        for token in chain_specific_tokens.iter() {
+            let _ = self
+                .chain_specific_token_ids(&universal_bridged_token_id)
+                .swap_remove(&token);
+
+            self.universal_bridged_token_pair(&token).clear();
+        }
+
+        self.universal_bridged_token_id()
```
Comment (dorin-iancu) [inline bridged-tokens-wrapper/src/lib.rs:49]: You already removed this at the start of this function.

**Review by dorin-iancu — CHANGES_REQUESTED** [review]

File: bridged-tokens-wrapper/src/lib.rs (hunk 1 above)
Comment (alyn509) [reply bridged-tokens-wrapper/src/lib.rs:49]: nope.. it is a separate storage.. it should get renamed

File: bridged-tokens-wrapper/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+#[elrond_wasm::contract]
+pub trait BridgedTokensWrapper {
+    #[init]
+    fn init(
+        &self,
+        universal_bridged_token_id: TokenIdentifier,
+        #[var_args] chain_specific_tokens: MultiValueEncoded<TokenIdentifier>,
+    ) {
+        self.universal_bridged_token_id()
+            .insert(universal_bridged_token_id.clone());
```
Comment (claudiulataretu) [inline bridged-tokens-wrapper/src/lib.rs:16]: might be better to use `set_if_empty` or `set`

File: bridged-tokens-wrapper/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+        for token_id in chain_specific_tokens {
+            let _ = self
+                .chain_specific_token_ids(&universal_bridged_token_id)
+                .insert(token_id);
+        }
+    }
+
+    #[only_owner]
+    #[endpoint(addWrappedToken)]
```
Comment (claudiulataretu) [inline bridged-tokens-wrapper/src/lib.rs:26]: I don't see a need for this endpoint if you are setting it already on init. This token should remain unchanged

File: bridged-tokens-wrapper/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+            new_payments.push(new_payment);
+        }
+
+        let caller = self.blockchain().get_caller();
+
+        self.send().direct_multi(&caller, &new_payments, &[]);
+
+        for payment in &new_payments {
+            self.token_liquidity(&payment.token_identifier)
```
Comment (claudiulataretu) [inline bridged-tokens-wrapper/src/lib.rs:158]: I would make this update in the first for. You might be able to save some gas.

File: multisig/interaction/snippets.sh [hunk 5]
```diff
@@ -255,9 +255,9 @@ setLocalRolesMultiTransferEsdt() {
...
 }
 
-setLocalRolesWrappedBridgedUSDC() {
-    getWrappedBridgedUSDCAddress
-    bech32ToHex ${getWrappedBridgedUSDCAddress}
+setLocalRolesBridgedTokensWrapper() {
+    getBridgedTokensWrapperAddress
+    bech32ToHex ${getBridgedTokensWrapperAddress}
 
     local LOCAL_BURN_ROLE=0x45534454526f6c654c6f63616c4275726e # "ESDTRoleLocalBurn"
```
Comment (claudiulataretu) [inline multisig/interaction/snippets.sh:262]: You can use this convertor from string to hex:
`"0x$(echo -n $2 | xxd -p -u | tr -d '\n')"`

File: bridged-tokens-wrapper/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            .chain_specific_to_universal_mapping(&payment_token)
+            .get();
+        let chain_specific_token_id = &payment_token;
+        require!(
+            self.chain_specific_token_ids(&universal_bridged_token_id)
+                .contains(chain_specific_token_id),
+            "Wrong esdt token"
+        );
+
+        require!(payment_amount > 0u32, "Payment must be more than 0");
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:100]: waiting for the next release - to delete these requires 🗡️

File: bridged-tokens-wrapper/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            .swap_remove(&chain_specific_token_id);
+
+        self.chain_specific_to_universal_mapping(&chain_specific_token_id)
+            .clear();
+    }
+
+    // endpoints
+
+    #[payable("*")]
+    #[endpoint(wrapToken)]
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:84]: we do not need the single endpoint - if we have the multiple.

File: bridged-tokens-wrapper/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+            self.token_liquidity(&payment.token_identifier)
+                .update(|value| *value += &payment.amount);
+        }
+
+        new_payments
+    }
+
+    #[payable("*")]
+    #[endpoint(unwrapToken)]
+    fn unwrap_token(
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:167]: we could do a multiple unwrap here as well - like the one above.

File: bridged-tokens-wrapper/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    fn whitelist_token(
+        &self,
+        chain_specific_token_id: TokenIdentifier,
+        universal_bridged_token_id: TokenIdentifier,
+    ) {
+        let _ = self
+            .chain_specific_token_ids(&universal_bridged_token_id)
+            .insert(chain_specific_token_id.clone());
+
+        let _ = self
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:61]: on whitelist token add the check for local burn and local mint

File: bridged-tokens-wrapper/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+        self.send().direct(
+            &caller,
+            &universal_bridged_token_id,
+            0,
+            &payment_amount,
+            &[],
+        );
+
+        self.token_liquidity(chain_specific_token_id)
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:115]: update of liquidity should be before send

File: bridged-tokens-wrapper/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            new_payments.push(new_payment);
+        }
+
+        let caller = self.blockchain().get_caller();
+
+        self.send().direct_multi(&caller, &new_payments, &[]);
+
+        for payment in &new_payments {
+            self.token_liquidity(&payment.token_identifier)
+                .update(|value| *value += &payment.amount);
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:159]: update of liquidity should be before send

File: bridged-tokens-wrapper/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+        self.send()
+            .esdt_local_burn(&universal_bridged_token_id, 0, &payment_amount);
+
+        let caller = self.blockchain().get_caller();
+
+        self.send()
+            .direct(&caller, chain_specific_token_id, 0, &payment_amount, &[]);
+
+        self.token_liquidity(chain_specific_token_id)
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:197]: update of liquidity should be before send

**Review by dorin-iancu — APPROVED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**